src/ray/raylet/raylet.cc

```diff
@@ -1,56 +1,129 @@
-    : acceptor_(io_service, boost::asio::local::stream_protocol::endpoint(socket_name)),
-      socket_(io_service),
-      tcp_acceptor_(io_service,
-                    boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0)),
-      tcp_socket_(io_service),
-      object_manager_(io_service, object_manager_config, gcs_client),
-      node_manager_(socket_name, resource_config, object_manager_),
-      gcs_client_(gcs_client) {
-  ClientID client_id = RegisterGcs();
-  object_manager_.SetClientID(client_id);
+               std::shared_ptr<gcs::AsyncGcsClient> gcs_client)
+    : acceptor_(main_service, boost::asio::local::stream_protocol::endpoint(socket_name)),
+      socket_(main_service),
+      object_manager_acceptor_(
+          main_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0)),
+      object_manager_socket_(main_service),
+      node_manager_acceptor_(
+          main_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 0)),
+      node_manager_socket_(main_service),
+      gcs_client_(gcs_client),
+      object_manager_(main_service, std::move(object_manager_service),
+                      object_manager_config, gcs_client),
+      node_manager_(main_service, node_manager_config, object_manager_, gcs_client_) {
   // Start listening for clients.
   DoAccept();
-  DoAcceptTcp();
+  DoAcceptObjectManager();
+  DoAcceptNodeManager();
+
+  RAY_CHECK_OK(RegisterGcs(main_service, node_manager_config));
+
+  RAY_CHECK_OK(RegisterPeriodicTimer(main_service));
+}
+
+Raylet::~Raylet() {
+  (void)gcs_client_->client_table().Disconnect();
+  RAY_CHECK_OK(object_manager_.Terminate());
+}
+
+ray::Status Raylet::RegisterPeriodicTimer(boost::asio::io_service &io_service) {
+  boost::posix_time::milliseconds timer_period_ms(100);
+  boost::asio::deadline_timer timer(io_service, timer_period_ms);
+  return ray::Status::OK();
 }
 
-Raylet::~Raylet() { RAY_CHECK_OK(object_manager_.Terminate()); }
+ray::Status Raylet::RegisterGcs(boost::asio::io_service &io_service,
+                                const NodeManagerConfig &node_manager_config) {
+  // TODO(hme): Clean up constants.
+  RAY_RETURN_NOT_OK(gcs_client_->Connect("127.0.0.1", 6379));
+  RAY_RETURN_NOT_OK(gcs_client_->Attach(io_service));
+
+  ClientTableDataT client_info = gcs_client_->client_table().GetLocalClient();
+  client_info.node_manager_address =
+      node_manager_acceptor_.local_endpoint().address().to_string();
+  client_info.object_manager_port = object_manager_acceptor_.local_endpoint().port();
+  client_info.node_manager_port = node_manager_acceptor_.local_endpoint().port();
+  // Add resource information.
+  for (const auto &resource_pair : node_manager_config.resource_config.GetResourceMap()) {
+    client_info.resources_total_label.push_back(resource_pair.first);
+    client_info.resources_total_capacity.push_back(resource_pair.second);
+  }
+
+  RAY_LOG(DEBUG) << "NM LISTENING ON: IP " << client_info.node_manager_address.c_str()
```

Comment:
Shouldn't need to do `.c_str()`, just pass in `client_info.node_manager_address`.